k8s/reconcilers/deployment.go

```diff
@@ -144,46 +144,108 @@ func (r *DeploymentReconciler) populateDiagnosticsAndActivitiesFromAnnotations(c
 }
 
 // attemptUpdate attempts to update the instance
-func (r *DeploymentReconciler) AttemptUpdate(ctx context.Context, object Reconcilable, log logr.Logger, operationStartTimeKey string, operationName string) (metrics.OperationStatus, reconcile.Result, error) {
+func (r *DeploymentReconciler) AttemptUpdate(ctx context.Context, object Reconcilable, isRemoval bool, log logr.Logger, operationStartTimeKey string, operationName string) (metrics.OperationStatus, reconcile.Result, error) {
 	// populate diagnostics and activities from annotations
 	ctx = r.populateDiagnosticsAndActivitiesFromAnnotations(ctx, object, operationName, r.kubeClient, log)
-	if !controllerutil.ContainsFinalizer(object, r.finalizerName) {
+	if !controllerutil.ContainsFinalizer(object, r.finalizerName) && !isRemoval {
 		controllerutil.AddFinalizer(object, r.finalizerName)
 		// updates the object in Kubernetes cluster
 		if err := r.kubeClient.Update(ctx, object); err != nil {
 			diagnostic.ErrorWithCtx(log, ctx, err, "failed to add finalizer to object")
 			return metrics.StatusUpdateFailed, ctrl.Result{}, err
 		}
 	}
+	// Get reconciliation interval
+	reconciliationInterval, timeout := r.deriveReconcileInterval(log, ctx, object)
 
-	if object.GetAnnotations()[operationStartTimeKey] == "" || utilsmodel.IsTerminalState(object.GetStatus().ProvisioningStatus.Status) {
-		r.patchOperationStartTime(object, operationStartTimeKey)
-		if err := r.kubeClient.Update(ctx, object); err != nil {
-			diagnostic.ErrorWithCtx(log, ctx, err, "failed to update object with operation start time")
+	if isRemoval {
+		// Timeout will be deletion timestamp + delete timeout duration
+		timeout := object.GetDeletionTimestamp().Time.Add(r.deleteTimeOut)
+
+		if metav1.Now().Time.After(timeout) {
+			return metrics.DeploymentTimedOut, ctrl.Result{RequeueAfter: reconciliationInterval}, nil
```

Comment:
We changed timeout behavior when deletion timeout to avoid the resources leaked. But this will block helm chart uninstall. Do you have corresponding helm chart changes?